Sources/Utils.php

```diff
@@ -998,6 +1004,78 @@ public static function ucwords(string $string): string
 		return self::convertCase($string, 'ucwords');
 	}
 
+	/**
+	 * Formats a string for use as a URL slug for queryless URLs.
+	 *
+	 * @param string $string The original string.
+	 * @param int $max_length Maximum length of the slug string. Default: 30.
+	 * @return string The slug string.
+	 */
+	public static function makeSlug(string $string, int $max_length = 30): string
+	{
+		static $common_words_regex;
+
+		// Decode any percent-encoding.
+		$string = rawurldecode($string);
+
+		// Decode any HTML entities.
+		$string = self::entityDecode($string, true);
+
+		// Convert to lower case and compatibility decomposition form.
+		$string = self::convertCase($string, 'lower', false, 'kd', true);
+
+		// Remove common words.
+		if (!isset($common_words_regex)) {
+			Lang::load('Search');
+
+			$common_words_regex =
+				'/\b' .
+				self::buildRegex(
+					array_map(
+						fn($word) => self::convertCase($word, 'lower', false, 'kd', true),
+						explode(',', Lang::getTxt('search_stopwords')),
+					),
+					'/',
+				) .
+				'\b/u';
+		}
+
+		$string = preg_replace($common_words_regex, ' ', $string);
+
+		// Build the slug.
+		$string = preg_replace_callback_array(
+			[
+				// Remove apostrophes, quotation marks, etc., within words.
+				'/(?<=[\p{L}\p{M}\p{N}])["\'\p{Pi}\p{Pf}](?=[\p{L}\p{M}\p{N}])/u' => fn($matches) => '',
+
+				// Remove combining and enclosing marks.
+				'/[\p{Mn}\p{Me}]/u' => fn($matches) => '',
+
+				// Replace all characters that aren't letters or numbers with hyphens.
+				// Note: ideographic characters count as letters.
+				'/[^\p{L}\p{N}]/u' => fn($matches) => '-',
+
+				// Collapse runs of hyphens.
+				'/-+/u' => fn($matches) => '-',
+
+				// Trim leading and trailing hyphens.
+				'/^-|-$/u' => fn($matches) => '',
+			],
+			// Remove all formatting characters, variation selectors, etc.
+			self::sanitizeChars($string, 2, ''),
+		);
+
+		// Convert back to composed form.
+		$string = self::normalize($string, 'c');
+
+		// Truncate to fit within the allowed length.
+		while (mb_strlen($string) > $max_length) {
+			$string = mb_substr($string, 0, str_contains($string, '-') ? mb_strrpos($string, '-') : $max_length);
+		}
+
```

Comment:
What if we add a special hook of this kind here?

```php
// Give mods a chance to change slug string.
IntegrationHook::call('integrate_make_slug', [&$string]);
```

Reply:
Good idea.